src/eos-util.c

```diff
@@ -0,0 +1,423 @@
+ *
+ * Authors: Vivek Dasmohapatra <vivek@etla.org>
+ *          Krzesimir Nowak <krzesimir@kinvolk.io>
+ */
+
+#include "eos-util.h"
+
+#include <libsoup/soup.h>
+
+#include <string.h>
+
+static const GDBusErrorEntry eos_updater_error_entries[] = {
+  { EOS_UPDATER_ERROR_WRONG_STATE, "com.endlessm.Updater.Error.WrongState" },
+  { EOS_UPDATER_ERROR_LAN_DISCOVERY_ERROR, "com.endlessm.Updater.Error.LANDiscoveryError" },
+  { EOS_UPDATER_ERROR_WRONG_CONFIGURATION, "com.endlessm.Updater.Error.WrongConfiguration" }
+};
+
+/* Ensure that every error code has an associated D-Bus error name */
+G_STATIC_ASSERT (G_N_ELEMENTS (eos_updater_error_entries) == EOS_UPDATER_N_ERRORS);
+
+GQuark
+eos_updater_error_quark (void)
+{
+  static volatile gsize quark_volatile = 0;
+  g_dbus_error_register_error_domain ("eos-updater-error-quark",
+                                      &quark_volatile,
+                                      eos_updater_error_entries,
+                                      G_N_ELEMENTS (eos_updater_error_entries));
+  return (GQuark) quark_volatile;
+}
+
+static const gchar * state_str[] = {
+   "None",
+   "Ready",
+   "Error",
+   "Polling",
+   "UpdateAvailable",
+   "Fetching",
+   "UpdateReady",
+   "ApplyUpdate",
+   "UpdateApplied" };
+
+G_STATIC_ASSERT (G_N_ELEMENTS (state_str) == EOS_UPDATER_N_STATES);
+
+const gchar *
+eos_updater_state_to_string (EosUpdaterState state)
+{
+  g_assert (state < EOS_UPDATER_N_STATES);
+
+  return state_str[state];
+};
+
+OstreeRepo *
+eos_updater_local_repo (void)
+{
+  GError *error = NULL;
+  g_autoptr(OstreeRepo) repo = ostree_repo_new_default ();
+
+  if (!ostree_repo_open (repo, NULL, &error))
+    {
+      GFile *file = ostree_repo_get_path (repo);
+      g_autofree gchar *path = g_file_get_path (file);
+
+      g_warning ("Repo at '%s' is not Ok (%s)",
+                 path ? path : "", error->message);
+
+      g_clear_error (&error);
+      g_assert_not_reached ();
+    }
+
+  return g_steal_pointer (&repo);
+}
+
+static gboolean
+ensure_is_ancestor (GFile *dir,
+                    GFile *file,
+                    GError **error)
+{
+  g_autoptr(GFile) child = NULL;
+
+  child = g_object_ref (file);
+  for (;;)
+    {
+      g_autoptr(GFile) parent = g_file_get_parent (child);
+
+      if (parent == NULL)
+        {
+          g_autofree gchar *raw_dir_path = g_file_get_path (dir);
+          g_autofree gchar *raw_file_path = g_file_get_path (file);
+
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                       "%s is not an ancestor of %s, not deleting anything",
+                       raw_dir_path,
+                       raw_file_path);
+          return FALSE;
+        }
+
+      if (g_file_equal (dir, parent))
+        break;
+
+      g_set_object (&child, parent);
+    }
+
+  return TRUE;
+}
+
+/* pass /a as the dir parameter and /a/b/c/d as the file parameter,
+ * will delete the /a/b/c/d file then /a/b/c and /a/b directories if
+ * empty.
+ */
```

Comment:
I suppose that makes sense. Is there any reason why we only clear out the parents if they're empty?

The convenience is certainly nice, but "delete this file" and "delete all parent directories where empty" are operations which in my mind are quite distinct and the name `delete_file_and_subdirs` makes me think of `rm -rf my_directory`.

Maybe `delete_file_and_empty_parents`?